src/compiler/semantic/analyser.rs

```diff
@@ -282,9 +309,43 @@ impl Analyser {
     }
 
     fn analyse_expression(&mut self, expr: Expression<BigInt, Identifier>) {
+        self.extract_usages_recursively(&expr);
         RULES.apply_expression(self, &expr)
     }
 
+    fn extract_id_usages(&mut self, ids: &[Identifier]) {
+        for id in ids {
+            self.symbols
+                .update_usages(&self.cur_scope, id.name(), id.debug_sym_ref());
+        }
+    }
+
+    fn extract_usages_recursively(&mut self, expr: &Expression<BigInt, Identifier>) {
+        match expr.clone() {
+            Expression::Query(dsym_ref, id) => {
+                self.symbols
+                    .update_usages(&self.cur_scope, id.name(), dsym_ref);
```

Comment:
@alxkzmn why in this case use the dsym of query and on others of the identifier itself?

Reply:
Let's take a look at how the parser is treating the different cases:

- expressions like `a + b`: `a` and `b` are eventually the queries to the signals so we can use their dsym_ref;
- assignments like `a <-- b + c` - although for a programmer it looks like both left and right side are similar to expressions above, for the parser the left side is an `ID` and the right side is an expression, so the right side will be recursively parsed as `Queries` above, but the left side needs special handling as `ID`s. Same goes for the state transitions like `-> final` - `final` here is an ID.

Reply:
Yes, what I was thinking is that to make the code look the same perhaps we could use the dsym of the identifier instead of the query. That in this case are both the same. But it is not important

Reply:
@leolara I think I got what you mean, we could refactor the `Query` to not have a separate debug sym ref because it can use the debug sym ref of its `Identifier`.

Reply:
The only obstacle for that is that in `Expression<F, V>` V is unbound. What's the motivation behind making V generic? Can it be anything besides `Identifier`?